x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/integrity/RepositoryIntegrityVerifier.java

```diff
@@ -0,0 +1,949 @@
+                                }
+                            })
+                            .addListener(listeners.acquire());
+                    }
+                }
+            }
+
+            /**
+             * Verifies the integrity of a specific shard snapshot
+             */
+            private class ShardSnapshotVerifier {
+                private final ShardContainerContents shardContainerContents;
+                private volatile boolean isRestorable = true;
+
+                ShardSnapshotVerifier(ShardContainerContents shardContainerContents) {
+                    this.shardContainerContents = shardContainerContents;
+                }
+
+                void run(ActionListener<Void> listener) {
+                    metadataTaskRunner.run(ActionRunnable.wrap(listener, this::verifyShardSnapshot));
+                }
+
+                private void verifyShardSnapshot(ActionListener<Void> listener) {
+                    final var shardId = shardContainerContents.shardId();
+                    final BlobStoreIndexShardSnapshot blobStoreIndexShardSnapshot;
+                    try {
+                        blobStoreIndexShardSnapshot = blobStoreRepository.loadShardSnapshot(
+                            blobStoreRepository.shardContainer(indexId, shardId),
+                            snapshotId
+                        );
+                    } catch (Exception e) {
+                        anomaly("failed to load shard snapshot").snapshotId(snapshotId)
+                            .shardDescription(indexDescription, shardId)
+                            .exception(e)
+                            .write(listener);
+                        return;
+                    }
+
+                    final var listeners = new RefCountingListener(1, listener.map(v -> {
+                        if (isRestorable) {
+                            restorableShardCount.incrementAndGet();
+                        }
+                        return v;
+                    }));
+                    final var shardGenerationConsistencyListener = listeners.acquire();
+
+                    runThrottled(
+                        Iterators.failFast(
+                            blobStoreIndexShardSnapshot.indexFiles().iterator(),
+                            () -> isCancelledSupplier.getAsBoolean() || listeners.isFailing()
+                        ),
+                        (releasable, fileInfo) -> verifyFileInfo(fileInfo, ActionListener.releaseAfter(listeners.acquire(), releasable)),
+                        1,
+                        blobsVerified,
+                        listeners
+                    );
+
+                    // NB this next step doesn't matter for restorability, it is just verifying that the shard gen blob matches the shard
+                    // snapshot blob
+                    verifyShardGenerationConsistency(blobStoreIndexShardSnapshot, shardGenerationConsistencyListener);
```

Comment:
Is it worth to treat these anomalies differently and not reporting `fail` for them? Does not have to be part of this PR anyway.

Reply:
No, it's still an integrity problem, just not one that affects restorability.